text/0000-export-executable-symbols.md

```diff
@@ -0,0 +1,116 @@
+- Feature Name: export-executable-symbols
+- Start Date: 2019-12-28
+- RFC PR: [rust-lang/rfcs#2841](https://github.com/rust-lang/rfcs/pull/2841)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add the ability to export symbols from executables, not just dylibs, via a new
+compiler flag: `-C export-executable-symbols`.
+
+# Motivation
+[motivation]: #motivation
+
+Java and C# can't statically link against C/Rust code.  Both require dylib
+symbols for their common native interop solution.  Which is fine if you let
+their executables call your dylib, but is a problem if you want your Rust
+executable to load a JVM instance, and let it call back into your executable.
+You might want to do this to allow you to:
+* Load multiple language runtimes into the same process (Rust + C# + Java + Lua anyone?  Only one of them can be the entry executable...)
+* Display user-friendly error messages if language runtimes are missing (maybe even a download link!)
+* [#[test] Java/Rust interop via cargo test.](https://github.com/MaulingMonkey/jerk/blob/04250c9d1b6ccc292eb27663f70919345c31007f/example-hello-world-jar/src/Global.rs)
+
+For this last case, I
+[manually export](https://github.com/MaulingMonkey/jerk/blob/04250c9d1b6ccc292eb27663f70919345c31007f/example-hello-world-jar/exports.def)
+executable symbols via
+[LINK](https://github.com/MaulingMonkey/jerk/blob/04250c9d1b6ccc292eb27663f70919345c31007f/example-hello-world-jar/build.rs#L4).
+This is ugly, brittle, and rustc
+[already knows](https://github.com/rust-lang/rust/blob/a916ac22b9f7f1f0f7aba0a41a789b3ecd765018/src/librustc_codegen_ssa/back/linker.rs#L706-L717)
+how to do this automatically, across more platforms, and better.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+https://doc.rust-lang.org/rustc/codegen-options/index.html could gain:
+
+```md
+## export-executable-symbols
+
+This flag causes `rustc` to export symbols from executables, as if they were dynamic libraries.
+
+You might use this to allow the JVM or MSCLR to call back into your executable's
+Rust code from Java/C# when embedding their runtimes into your Rust executable.
+```
+
+`rustc -C help` could gain:
+
+```
+    -C    export-executable-symbols -- export symbols from executables, as if they were dynamic libraries.
+```
+
+My Java interop [Quick Start](https://github.com/MaulingMonkey/jerk/blob/master/Readme.md#quick-start)
+would start recommending a `.cargo/config` with:
+```toml
+[build]
+rustflags = ["-C", "export-executable-symbols"]
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+On a technical level, this just involves preventing an early bailout when
```

Comment:
This is not a reference level explanation of the feature, but rather an approach to implementation. For RFC process this is not too useful and is prone to becoming invalid and impossible to understand in the future or in context of alternative implementations.

Instead this section should be exhaustively enumerating interactions with other *language* features (@pnkfelix mentioned for instance visibility) and describing how the relevant portions (symbols) of produced artifacts change as a result.